homeassistant/components/homekit/__init__.py

```diff
@@ -3,154 +3,197 @@
     SUPPORT_TARGET_TEMPERATURE_HIGH, SUPPORT_TARGET_TEMPERATURE_LOW)
+from homeassistant.const import (
+    ATTR_CODE, ATTR_SUPPORTED_FEATURES, ATTR_UNIT_OF_MEASUREMENT,
+    CONF_PORT, TEMP_CELSIUS, TEMP_FAHRENHEIT,
+    EVENT_HOMEASSISTANT_START, EVENT_HOMEASSISTANT_STOP)
+import homeassistant.helpers.config_validation as cv
+from homeassistant.helpers.entityfilter import FILTER_SCHEMA
 from homeassistant.util import get_local_ip
 from homeassistant.util.decorator import Registry
+from .const import (
+    DOMAIN, HOMEKIT_FILE, CONF_AUTO_START, CONF_ENTITY_CONFIG, CONF_FILTER,
+    DEFAULT_PORT, DEFAULT_AUTO_START, SERVICE_HOMEKIT_START)
+from .util import (
+    validate_entity_config, show_setup_message)
 
 TYPES = Registry()
 _LOGGER = logging.getLogger(__name__)
 
-_RE_VALID_PINCODE = r"^(\d{3}-\d{2}-\d{3})$"
-
-DOMAIN = 'homekit'
-REQUIREMENTS = ['HAP-python==1.1.7']
-
-BRIDGE_NAME = 'Home Assistant'
-CONF_PIN_CODE = 'pincode'
-
-HOMEKIT_FILE = '.homekit.state'
-
-
-def valid_pin(value):
-    """Validate pin code value."""
-    match = re.match(_RE_VALID_PINCODE, str(value).strip())
-    if not match:
-        raise vol.Invalid("Pin must be in the format: '123-45-678'")
-    return match.group(0)
+REQUIREMENTS = ['HAP-python==1.1.7', 'pypng==0.0.18']
 
 
 CONFIG_SCHEMA = vol.Schema({
     DOMAIN: vol.All({
-        vol.Optional(CONF_PORT, default=51826): vol.Coerce(int),
-        vol.Optional(CONF_PIN_CODE, default='123-45-678'): valid_pin,
+        vol.Optional(CONF_PORT, default=DEFAULT_PORT): vol.Coerce(int),
+        vol.Optional(CONF_AUTO_START, default=DEFAULT_AUTO_START): cv.boolean,
+        vol.Optional(CONF_FILTER, default={}): FILTER_SCHEMA,
+        vol.Optional(CONF_ENTITY_CONFIG, default={}): validate_entity_config,
     })
 }, extra=vol.ALLOW_EXTRA)
 
 
-@asyncio.coroutine
-def async_setup(hass, config):
+async def async_setup(hass, config):
     """Setup the HomeKit component."""
-    _LOGGER.debug("Begin setup HomeKit")
+    _LOGGER.debug('Begin setup HomeKit')
 
     conf = config[DOMAIN]
-    port = conf.get(CONF_PORT)
-    pin = str.encode(conf.get(CONF_PIN_CODE))
+    port = conf[CONF_PORT]
+    auto_start = conf[CONF_AUTO_START]
+    entity_filter = conf[CONF_FILTER]
+    entity_config = conf[CONF_ENTITY_CONFIG]
 
-    homekit = HomeKit(hass, port)
-    homekit.setup_bridge(pin)
+    homekit = HomeKit(hass, port, entity_filter, entity_config)
+    homekit.setup()
 
-    hass.bus.async_listen_once(
-        EVENT_HOMEASSISTANT_START, homekit.start_driver)
-    return True
+    if auto_start:
+        hass.bus.async_listen_once(EVENT_HOMEASSISTANT_START, homekit.start)
+        return True
+
+    def handle_homekit_service_start(service):
+        """Handle start HomeKit service call."""
+        if homekit.started:
+            _LOGGER.warning('HomeKit is already running')
+            return
+        homekit.start()
 
+    hass.services.async_register(DOMAIN, SERVICE_HOMEKIT_START,
```

Comment:
Should we only register this if autostart is disabled?

Reply:
I already do so, see L56-58.